utils/mil_ros_tools/online_bagger/nodes/online_bagger.py

```diff
@@ -0,0 +1,162 @@
+        self.streaming = True
+        self.dumping = False
+        # self.resubscribe
+
+        self.get_params()
+        self.make_dicts()
+        self.subscribe()
+
+
+        self.bagging_service = rospy.Service('/online_bagger', SetBool, self.start_bagging)
+
+    def get_params(self):
+
+        self.topics = rospy.get_param('/online_bagger/message_topics')
+        self.ram_limit = rospy.get_param('/online_bagger/ram_limit')
+        self.stream_time = rospy.get_param('/online_bagger/stream_time')
+
+        rospy.loginfo('stream_time: %s', self.stream_time)
+        # rospy.loginfo('ram_limit: %s', self.ram_limit)
+
+    def make_dicts(self):
+        '''
+        make dictionaries with deques() that will be filled with topics
+        with all of the 5topics defined by the yaml
+        '''
+
+        self.topic_list = {}
+ 
+        for topic in self.topics:
+            self.topic_list[topic] = deque()
+
+        rospy.loginfo('topics: %s', self.topic_list.keys())
+    def subscribe(self):
+        '''
+        Immediately initiated with instance of the BagStream class:
+        subscribes to the set of topics defined in the yaml configuration file
+        '''
+        # use failed topics list that can be resubscribed to at a later time?
+
+        for topic in self.topic_list.keys():
+            msg_class = rostopic.get_topic_class(topic)
+            if msg_class[1] is None:
+                self.failed_topics.append(topic)
+            else:
+                rospy.Subscriber(topic, msg_class[0], lambda msg, _topic=topic: self.callback(msg, _topic))
+
+        rospy.loginfo('failed topics: %s', self.failed_topics)
+
+    def get_oldest_topic_time(self, topic):
+        '''
+        returns oldest timestamp for a given topic
+        '''
+        return self.topic_list[topic][0][0]
+
+    def get_newest_topic_time(self, topic):
+        '''
+        returns newest timestamp for a given topic
+        '''
+        return self.topic_list[topic][-1][0]
+
+
+    def callback(self, msg, topic):
+        '''
+        streaming callback function, stops streaming during bagging process
+        also pops off msgs from dequeue if length of stream is longer than
+        stream time specified in yaml parameters
+
+        stores all data in a dictionary.
+        the keys are topic names
+        the values are dequeues of tuples
+        each tuple is a time stamp of rostime and the message for that topic
+
+        if any errors are occuring you can uncomment the lines with
+        if self.n % 50  == 0 to display every 50th message, time diff and topic
+        this will not give every 50th message on each topic, it provides every
+        50th message accross all topics.
+        
+        stream, callback function does nothing if streaming is not active
+        '''
+
+        if not(self.streaming):
+            return
+
+        self.n = self.n + 1
+        self.topic_list[topic].append((rospy.get_rostime(), msg))
```

Comment:
Might be a good idea to check if the message has a header and use that as the time if it exists. ```python
if hasattr(msg, 'header'):
    time = msg.header.stamp
else:
    time = rospy.get_rostime()
```